src/vec_int128_ppc.h

```diff
@@ -2388,73 +2889,366 @@ vec_mulluq (vui128_t a, vui128_t b)
- *  |power9   | 16    | 2/cycle  |
+ *  |power8   | 56-64 | 1/cycle  |
+ *  |power9   | 33-39 | 1/cycle  |
  *
- *  @param vra a 128-bit vector treated a __int128.
- *  @return a 128-bit vector with bits 121:127 containing the
- *  population count.
+ *  @param *mulu pointer to upper 128-bits of the product.
+ *  @param a 128-bit vector treated a __int128.
+ *  @param b 128-bit vector treated a __int128.
+ *  @return vector unsigned __int128 (lower 128-bits) of a * b.
  */
 static inline vui128_t
-vec_popcntq (vui128_t vra)
+vec_muludq (vui128_t *mulu, vui128_t a, vui128_t b)
 {
-  vui64_t result;
-
+  vui32_t t, tmq;
+  /* compute the 256 bit product of two 128 bit values a, b.
+   * The high 128 bits are accumulated in t and the low 128-bits
+   * in tmq. The high 128-bits of the product are returned to the
+   * address of the 1st parm. The low 128-bits are the return
+   * value.
+   */
+#ifdef _ARCH_PWR9
+  const vui64_t zero = { 0, 0 };
+  vui64_t b_swap = vec_swapd ((vui64_t) b);
+  vui128_t tmh, tab, tba, tb0, tc1, tc2;
+  /* multiply the low 64-bits of a and b.  For PWR9 this is just
+   * vmsumudm with conditioned inputs.  */
+  tmq = (vui32_t) vec_vmuloud ((vui64_t)a, (vui64_t)b);
+  /* compute the 2 middle partial projects.  Can't directly use
+   * vmsumudm here because the sum of partial products can overflow.  */
+  tab = vec_vmuloud ((vui64_t) a, b_swap);
+  tba = vec_vmuleud ((vui64_t) a, b_swap);
+  t   = (vui32_t) vec_adduqm (tab, tba);
+  tc1 = vec_addcuq (tab, tba);
+  tmh = (vui128_t) vec_mrgahd ((vui128_t) zero, (vui128_t) tmq);
+  t   = (vui32_t) vec_adduqm ((vui128_t) t, tmh);
+  tc2 = vec_addcuq ((vui128_t) t, tmh);
+  tc1 = (vui128_t) vec_vadduwm ((vui32_t) tc1, (vui32_t) tc2);
+  /* result = t[l] || tmq[l].  */
+  tmq = (vui32_t) vec_mrgald ((vui128_t) t, (vui128_t) tmq);
+  /* we can use multiply sum here because the high product plus the
+   * high sum of middle partial products can't overflow.  */
+  t   = (vui32_t) vec_permdi ((vui64_t) tc1, (vui64_t) t, 2);
+  tb0 = (vui128_t) vec_mrgahd ((vui128_t) b, (vui128_t) zero);
+  /* sum = (a[h] * b[h]) + (a[l] * 0) + (tc1[l] || t[h]).  */
+  t   = (vui32_t) vec_msumudm ((vui64_t) a, (vui64_t) tb0, (vui128_t) t);
+#else
 #ifdef _ARCH_PWR8
-  /*
-   * Use the Vector Population Count Doubleword instruction to get
-   * the count for the left and right vector halves.  Then sum across
-   * the left and right counts to get the final 128-bit vector count
-   * (0-128).
+  vui32_t tsw;
+  vui32_t tc;
+  vui32_t t_odd, t_even;
+  vui32_t z = { 0, 0, 0, 0 };
+  /* We use the Vector Multiple Even/Odd Unsigned word to compute
+   * the 128 x 32 partial (160-bit) product of value a with the
+   * word splat of b. These instructions (vmuleum, vmuloum)
+   * product four 64-bit 32 x 32 partial products where even
+   * results are shifted 32-bit left from odd results. After
+   * shifting the high 128 bits can be summed via Vector add
+   * unsigned quadword.
```

Comment:
"results are shifted 32 bit*s* left from odd results".  I do not understand this phrase.
```
After shifting<comma> the high 128 bits can be summed via Vector *A*dd *U*nsigned *Q*uadword.
```
?

Reply:
Ok we will try this wording:

We use the Vector Multiple Even/Odd Unsigned Word to compute
 the 128 x 32 partial (160-bit) product of value a with the
 word splat of b. This produces four 64-bit (32 x 32)
 partial products in two vector registers. These results
 are not aligned for summation as is. So the odd result is
 shifted right 32-bits before it is summed (via Vector Add
 Unsigned Quadword Modulo) with the the even result.
 The low order 32-bits, of the 160-bit product
 is shifted (right) in to a separate vector (tmq).

 This is repeated for each (low to high order) word of b.
 After the first (160-bit) partial product, the high 128-bits
 (t) of the previous partial product is summed with the current
 odd multiply result, before this sum (including any carry out)
 is shifted right 32-bits.  Bits shifted out of the of this sum
 are shifted (32-bits at a time) into the low order 128-bits
 of the product (tmq). The shifted odd sum is then added to the
 current even product, After the 4th step this sum is the
 final high order 128-bits of the quadword product.

I plan to include this full test in the _ARCH_PWR8 section of vec_muludq. muhluq and mulluq
Will include a comment referring to this. Something like:


  We use Vector Multiply Even/Odd Unsigned Word to compute
 the 128 x 32 partial (160-bit) product of vector a with a
 word element of b. The (for each word of vector b) 4 X 160-bit
 partial products are  summed to produce the full 256-bit product.
 See the comment in vec_muludq for details.